File: api/models/boarding_house.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type BoardingHouse struct {
	ID            uuid.UUID `gorm:"type:char(36);primaryKey" json:"id"`
	OwnerID       uuid.UUID `gorm:"type:char(36);not null" json:"owner_id"`
	Name          string    `gorm:"type:varchar(255);not null" json:"name"`
	Description   string    `gorm:"type:text" json:"description"`
	Address       string    `gorm:"type:text" json:"address"`
	City          string    `gorm:"type:varchar(100)" json:"city"`
	PricePerMonth float64   `gorm:"type:decimal(10,2)" json:"price_per_month"`
	RoomAvailable int       `gorm:"default:0" json:"room_available"`
	GenderAllowed string    `gorm:"type:enum('male','female','mixed')" json:"gender_allowed"`

	Owner      User                 `gorm:"foreignKey:OwnerID" json:"owner"`
	Facilities []Facility           `gorm:"many2many:boarding_house_facilities;" json:"facilities"`
	Images     []BoardingHouseImage `gorm:"foreignKey:BoardingHouseID" json:"images"`

	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}

func (item *BoardingHouse) BeforeCreate(tx *gorm.DB) (err error) {
	if item.ID == uuid.Nil {
		item.ID = uuid.New()
	}
	return
}

// BoardingHouseRequest represents the request payload for creating or updating a BoardingHouse
type BoardingHouseRequest struct {
	OwnerID       string  `json:"owner_id" binding:"required,uuid"`
	Name          string  `json:"name" binding:"required"`
	Description   string  `json:"description"`
	Address       string  `json:"address" binding:"required"`
	City          string  `json:"city" binding:"required"`
	PricePerMonth float64 `json:"price_per_month" binding:"required,gt=0"`
	RoomAvailable int     `json:"room_available" binding:"required,gte=0"`
	GenderAllowed string  `json:"gender_allowed" binding:"required,oneof=male female mixed"`
}

// BoardingHouseResponse represents the response payload for a BoardingHouse
type BoardingHouseResponse struct {
	ID            string                       `json:"id"`
	OwnerID       string                       `json:"owner_id"`
	Name          string                       `json:"name"`
	Description   string                       `json:"description"`
	Address       string                       `json:"address"`
	City          string                       `json:"city"`
	PricePerMonth float64                      `json:"price_per_month"`
	RoomAvailable int                          `json:"room_available"`
	GenderAllowed string                       `json:"gender_allowed"`
	Facilities    []FacilityResponse           `json:"facilities"`
	Images        []BoardingHouseImageResponse `json:"images"`
	CreatedAt     time.Time                    `json:"created_at"`
	UpdatedAt     time.Time                    `json:"updated_at"`
}

File: mobile_application/lib/models/boarding_house.dart
import 'package:mobile_application/models/user.dart';

class BoardingHouse {
  final String id;
  final String ownerId;
  final String name;
  final String description;
  final String address;
  final String city;
  final int pricePerMonth;
  final int roomAvailable;
  final String genderAllowed;
  final List<Facility> facilities;
  final List<BoardingHouseImage> images;
  final User owner;
  final int bookedCount;
  final int availableCount;

  BoardingHouse({
    required this.id,
    required this.ownerId,
    required this.name,
    required this.description,
    required this.address,
    required this.city,
    required this.pricePerMonth,
    required this.roomAvailable,
    required this.genderAllowed,
    required this.facilities,
    required this.images,
    required this.bookedCount,
    required this.availableCount,
    required this.owner,
  });

  factory BoardingHouse.fromJson(Map<String, dynamic> json) {
    return BoardingHouse(
      id: json['id'] ?? '',
      ownerId: json['owner_id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      address: json['address'] ?? '',
      city: json['city'] ?? '',
      pricePerMonth: json['price_per_month'] ?? 0,
      roomAvailable: json['room_available'] ?? 0,
      genderAllowed: json['gender_allowed'] ?? '',
      facilities:
          (json['facilities'] as List<dynamic>?)
              ?.map((facilityJson) => Facility.fromJson(facilityJson))
              .toList() ??
          [],
      images:
          (json['images'] as List<dynamic>?)
              ?.map((imgJson) => BoardingHouseImage.fromJson(imgJson))
              .toList() ??
          [],
      bookedCount: json['booked_count'] ?? 0,
      availableCount: json['available_count'] ?? 0,
      owner: User.fromJson(json['owner'] ?? {}),
    );
  }

  static String getGenderLabelStatic(String gender) {
    const genderMapping = {
      'male': 'Putra',
      'female': 'Putri',
      'mixed': 'Campur',
    };
    return genderMapping[gender] ?? 'Tidak Diketahui';
  }
}

class Facility {
  final String id;
  final String name;

  Facility({required this.id, required this.name});

  factory Facility.fromJson(Map<String, dynamic> json) {
    return Facility(id: json['id'], name: json['name']);
  }
}

class BoardingHouseImage {
  final String id;
  final String boardingHouseId;
  final String imageUrl;
  final String createdAt;
  final String updatedAt;

  BoardingHouseImage({
    required this.id,
    required this.boardingHouseId,
    required this.imageUrl,
    required this.createdAt,
    required this.updatedAt,
  });

  factory BoardingHouseImage.fromJson(Map<String, dynamic> json) {
    return BoardingHouseImage(
      id: json['id'],
      boardingHouseId: json['boarding_house_id'],
      imageUrl: json['image_url'],
      createdAt: json['created_at'],
      updatedAt: json['updated_at'],
    );
  }
}
